### PR TITLE
General updates

### PR DESCRIPTION
**Pull request by JayH5**
* `gdbm-dev` now in base image
* pip 9.0.1
* Rework build script -- just do things in Travis file
* Rewrite deploy script in Python
* Run the official-images tests

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -15,16 +15,32 @@ before_install:
...
+  - git clone https://github.com/docker-library/official-images.git ~/official-images
+
+before_script:
+  - image="$(awk '$1 == "FROM" { print $2; exit }' $VERSION/onbuild/Dockerfile)${VARIANT:+-$VARIANT}"
+  - version="$(awk '$2 == "PYTHON_VERSION" { print $3; exit }' $VERSION/${VARIANT:+$VARIANT/}Dockerfile)"
+  - echo "Building image $image with Python version $version"
+
 script:
-  - ./build.sh
+  - docker build -t "$image" "$VERSION${VARIANT:+/$VARIANT}"
```
Comment (jerith) [inline .travis.yml]: I think this is the first time I've ever seen `:+` outside of the documentation. :-)

File: .travis.yml [hunk 2]
```diff
@@ -15,16 +15,32 @@ before_install:
...
 after_script:
   - docker images
 
 before_deploy:
-  - pip install docker-ci-deploy
+  - docker login -u "$REGISTRY_USER" -p "$REGISTRY_PASS"
 deploy:
   provider: script
-  script: ./deploy.sh
+  script: ./deploy.py "$image" "$version" && if [[ -z "$VARIANT" ]]; then ./deploy.sh "${image}-onbuild" "$version"; fi
```
Comment (jerith) [inline .travis.yml]: I'm not hugely enthusiastic about this very long command line. Can we maybe have `deploy.py` handle that somehow?

Also, are we actually using `$VARIANT`? I don't see anything that sets it.

**Review by jerith — COMMENTED** [review]
I think `deploy.py` is complex enough that some of its functions need tests. A single test module that exercises `generate_versioned_tags()` and friends should be sufficient.

File: .travis.yml (hunk 2 above)
Comment (JayH5) [reply .travis.yml]: Nope :/ it's mostly just there to keep things in common with the upstream official images. There it is used for the `slim` Debian variant but a `slim` Alpine image would be equivalent to the official `alpine` image. I guess I should just remove it...

Comment (jerith) [reply .travis.yml]: I think we should ditch it, then. Nothing's using the code (or likely to in the near future) so it's unnecessary complexity.

File: .travis.yml [hunk 3]
```diff
@@ -8,23 +8,39 @@ env:
...
 after_script:
   - docker images
 
 before_deploy:
-  - pip install docker-ci-deploy
+  - docker login -u "$REGISTRY_USER" -p "$REGISTRY_PASS"
 deploy:
   provider: script
-  script: ./deploy.sh
+  script: ./deploy.py "$image" "$version" && ./deploy.sh "${image}-onbuild" "$version"
```
Comment (jerith) [inline .travis.yml]: `deploy.py` followed by `deploy.sh`?

File: test_deploy.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    def test_other_version_tag(self):
+        versioned_tags = generate_versioned_tags('2.7-foo', '5.4.1')
+        self.assertEqual(versioned_tags,
+                         ['5.4.1-2.7-foo', '5.4-2.7-foo', '5-2.7-foo'])
+
+    def test_version_tag(self):
+        versioned_tags = generate_versioned_tags('5.4', '5.4.1')
+        self.assertEqual(versioned_tags, ['5.4.1', '5.4', '5'])
+
+    def test_nont_tag(self):
```
Comment (jerith) [inline test_deploy.py]: "nont" -> "none"?

**Review by jerith — APPROVED** [review]
:+1:

**Pull request merged**